### PR TITLE
Validate Asynchronously

### PR DESCRIPTION
**Pull request by lukehesluke**
Issue: https://github.com/openactive/data-model-validator/issues/314

🚨 Relies on this PR in order for tests to pass, so this one should be merged first: https://github.com/openactive/data-model-validator/pull/316 🚨

## Notes for reviewer

This is a large PR with thousands of lines changed. This was unfortunately necessary as there's no easy way to have the same code choose to be either synchronous or asynchronous in JavaScript.

Most of the lines are very simple one-line changes e.g. moving the tests to async. I've tried to signpost where I can big chunks of code that are just duplicates and therefore need minimal review.

------

The rationale for this can be found in the issue. Briefly, validator makes IO calls synchronously which could be very harmful to performance in applications running validator concurrently.

In order to maintain backwards compatibility, I've left the `validate` function as a synchronous function and created a new function, `validateAsync`.

Unfortunately, this has also meant having to split every function that has the potential to touch async into -Sync and -Async functions. This includes:

- `src/validate.js`
  - validate(Async/Sync)
  - ApplyRules.applyModelRules(Async/Sync)
  - ApplyRules.applySubModelRules(Async/Sync)
- `src/rules/rule.js`
  - validate(Async/Sync)
  - validateField(Async/Sync)
- `src/rules/raw-rule.js` (these rules never touch async code but they were split so as to preserve consistency with parent class Rule)
  - validate(Async/Sync)
  - validateRaw(Async/Sync)
- `src/rules/core/fields-not-in-model-rule.js`
  - getContexts(Async/Sync)
- `src/helpers/json-loader.js`
  - `getFile(Async/Sync)`

--------

Progress:

- [x] Split application into -Sync and -Async
  - [x] json-loader:
    - [x] sync-request (via request)
    - [x] fs.existsSync
    - [x] fs.readFileSync
  - [x] Rules
  - [x] Raw Rules
  - [x] validate.js
- [x] Update README
- [x] Update tests
- [x] Update docs

### CODE REVIEW

File: src/rules/core/fields-not-in-model-rule.js [hunk 1]
```diff
@@ -251,7 +256,269 @@ module.exports = class FieldsNotInModelRule extends Rule {
     };
   }
 
-  validateField(node, field) {
+  async getContextsAsync(node, field) {
```
Comment (lukehesluke) [inline src/rules/core/fields-not-in-model-rule.js]: This is just a copy of `getContextsSync` but with  `await` and `-Async` functions

File: src/rules/core/fields-not-in-model-rule.js [hunk 2]
```diff
@@ -251,7 +256,269 @@ module.exports = class FieldsNotInModelRule extends Rule {
...
+      errors,
+    };
+  }
+
+  /**
+   * @deprecated since version 1.2.0, since it uses synchronous IO
+   *   (https://nodejs.org/en/docs/guides/blocking-vs-non-blocking/#comparing-code). Use
+   *   validateFieldAsync() instead
+   */
+  validateFieldSync(node, field) {
```
Comment (lukehesluke) [inline src/rules/core/fields-not-in-model-rule.js]: This method is the exact same as the previous `validateField` method - Git is just marking it as new as it's been moved around

File: src/rules/data-quality/activity-in-activity-list-rule.js [hunk 3]
```diff
@@ -92,7 +92,7 @@ module.exports = class ActivityInActivityListRule extends Rule {
     };
   }
 
-  validateField(node, field) {
+  async validateFieldAsync(node, field) {
```
Comment (lukehesluke) [inline src/rules/data-quality/activity-in-activity-list-rule.js]: This function is the same as `validateField` except the one line where it uses `await JsonLoaderHelper.getFileAsync`

File: src/rules/data-quality/activity-in-activity-list-rule.js [hunk 4]
```diff
@@ -165,7 +165,201 @@ module.exports = class ActivityInActivityListRule extends Rule {
...
+          }
+        }
+        index += 1;
+      }
+    }
+
+    return errors;
+  }
+
+  validateFieldSync(node, field) {
```
Comment (lukehesluke) [inline src/rules/data-quality/activity-in-activity-list-rule.js]: This function is identical to the previous `validateField` method

File: src/validate.js [hunk 5]
```diff
@@ -102,23 +107,214 @@ class ApplyRules {
...
       if (Object.prototype.hasOwnProperty.call(nodeToTest.value, field)) {
         // If this field is itself a model, apply that model's rules to it
-        errors = errors.concat(this.applySubModelRules(rules, nodeToTest, field));
+        errors = errors.concat(this.applySubModelRulesSync(rules, nodeToTest, field));
       }
     }
     return errors;
   }
+
+  static async applySubModelRulesAsync(rules, nodeToTest, field) {
```
Comment (lukehesluke) [inline src/validate.js]: This is almost identical to the previous `applySubModelRules` method except uses `await` where appropriate

File: src/validate.js [hunk 6]
```diff
@@ -102,23 +107,214 @@ class ApplyRules {
...
+          );
+          errors = errors.concat(subModelErrors);
+        }
+      }
+      index += 1;
+    }
+    return errors;
+  }
+
+  static async applyModelRulesAsync(rules, nodeToTest) {
```
Comment (lukehesluke) [inline src/validate.js]: This is almost identical to the previous `applyModelRules` method except uses await where appropriate

File: src/validate.js [hunk 7]
```diff
@@ -102,23 +107,214 @@ class ApplyRules {
...
+        // If this field is itself a model, apply that model's rules to it
+        const subModelErrors = await ApplyRules.applySubModelRulesAsync(rules, nodeToTest, field);
+        errors = errors.concat(subModelErrors);
+      }
+    }
+    return errors;
+  }
+}
+
+async function validateAsync(value, options) {
```
Comment (lukehesluke) [inline src/validate.js]: This is almost identical to the previous `validate` method except uses await where appropriate

File: src/validate-spec.js [hunk 8]
```diff
@@ -197,370 +198,686 @@ describe('validate', () => {
...
     });
   });
+  afterEach(() => {
+    nock.cleanAll();
+    nock.enableNetConnect();
+  });
 
-  it('should return a failure if passed an invalid model', () => {
-    const data = {};
+  describe('validateAsync()', () => {
```
Comment (lukehesluke) [inline src/validate-spec.js]: There is also a `describe('validateSync()', () => {` describe block which is a duplicate of this block except that it tests `validateSync` instead

**Comment by nickevansuk** [comment]
Hey @lukehesluke, would you be able to help us out with your logic for implementing caching manually rather than via something from npm?

**Comment by lukehesluke** [comment]
Hi @nickevansuk. I think you're asking why I have implemented caching manually rather than via an NPM library.

The short answer is that I have matched the existing logic as much as I can (while adding an asynchronous path) so this question might be outside of the scope of this PR.

**Comment by nickevansuk** [comment]
Thanks @lukehesluke ok that makes sense, so you haven't added additional caching logic here, only replicated what existed already as an async version? (so the old stuff didn't use npm, nor does the new stuff?)

**Comment by lukehesluke** [comment]
Yes @nickevansuk . I haven't updated the caching mechanism at all in this PR except for adding a way to do the very same thing but asynchronously

**Pull request merged**